File: csharp/1590.make-sum-divisible-by-p.cs
/*
 * @lc app=leetcode id=1590 lang=csharp
 *
 * [1590] Make Sum Divisible by P
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int MinSubarray(int[] nums, int p) {
        int acc = 0, curr = 0, n = nums.Length, ret = n;
        var indices = new Dictionary<int, int>();
        foreach (var x in nums) {
            acc = (acc + x) % p;
        }
        indices[0] = -1;
        for (int i = 0; i < n; ++i) {
            curr  = (curr + nums[i]) % p;
            indices[curr] = i;
            if (indices.TryGetValue((-acc + p + curr) % p, out var j)) {
                ret = Math.Min(ret, i - j);
            }
        }
        return ret == n ? -1 : ret;
    }
}
// @lc code=end

File: java/1590.make-sum-divisible-by-p.java
/*
 * @lc app=leetcode id=1590 lang=java
 *
 * [1590] Make Sum Divisible by P
 */

import java.util.HashMap;
import java.util.Map;

// @lc code=start

class Solution {
    public int minSubarray(int[] nums, int p) {
        int n = nums.length, acc = 0;
        for (int x : nums) {
            acc = (acc + x) % p;
        }
        int curr = 0, ret = n;
        Map<Integer, Integer> lastIndex = new HashMap<>();
        lastIndex.put(0, -1);
        for (int i = 0; i < n; ++i) {
            curr = (curr + nums[i]) % p;
            lastIndex.put(curr, i);
            Integer l = lastIndex.get((curr - acc + p) % p);
            if (l == null) {
                continue;
            }
            ret = Math.min(ret, i - l);
        }
        return ret == n ? -1 : ret;
    }
}
// @lc code=end

